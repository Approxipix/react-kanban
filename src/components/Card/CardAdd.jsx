import React, { Component } from "react";
import PropTypes from 'prop-types';
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { addCard } from '../../redux/cardReducer/actions/actions';
import { TextArea, Actions, SubmitButton, CancelButton } from '../BaseComponent';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import ClickOutside from "../ClickOutside/ClickOutside";
import styled from 'styled-components';

const Wrapper = styled.div`
  padding: .5rem;
`;

const AddButton = styled.button`
  text-align: left;
  width: 100%;
  padding: .5rem .75rem;
  font-size: .875rem; 
  color: #6b778c;
  transition: all .2s ease-in;
  cursor: pointer;
  &:hover {
    color: #172b4d;
    background-color: rgba(9,30,66,.13);
  }
`;
AddButton.displayName = 'AddButton';

class CardAdd extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
      title: '',
    };
    this.textarea = React.createRef();
  }

  toggleOpened = () => {
    this.setState({
      isOpened: !this.state.isOpened,
      title: '',
    })
  };

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.toggleOpened();
    }
  };

  handleChange = (e) => {
    this.setState({
      title: e.target.value,
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    if (!title)  return;
    this.props.actions.addCard({
      listId: this.props.listId,
      cardTitle: title,
      newCardId: Date.now().toString(),
    });
    this.setState({ title: '' }, () => {
      if (!this.textarea.focus) return;
      this.textarea.focus()
    })
  };

  render = () => {
    const { isOpened, title } = this.state;

    return isOpened ? (
      <ClickOutside action={'mousedown'} toggleOpened={this.toggleOpened}>
        <Wrapper>
          <form onSubmit={this.handleSubmit}>
            <TextArea
              type="text"
              name="title"
              value={title}
              size=".875rem;"
              padding=".5rem"
              rows={3}
              placeholder="Enter a card title..."
              ref={(e) => { this.textarea = e }}
              onKeyDown={this.handleKeyDown}
              onChange={(e) => this.handleChange(e)}
              spellCheck={false}
              autoFocus
            />
            <Actions>
              <SubmitButton type="submit" disabled={!title}>
                Create card
              </SubmitButton>
              <CancelButton onClick={() => this.toggleOpened()}>
                <FontAwesomeIcon icon="times"/>
              </CancelButton>
            </Actions>
          </form>
        </Wrapper>
      </ClickOutside>
    ) : (
      <AddButton onClick={() => this.toggleOpened()}>
        + Add another card
      </AddButton>
    );
  };
}

CardAdd.propTypes = {
  listId: PropTypes.string.isRequired,
};

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addCard: addCard,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CardAdd);
