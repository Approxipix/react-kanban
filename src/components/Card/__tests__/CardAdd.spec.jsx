import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CardAdd from '../CardAdd';

const mockStore = configureStore([]);

describe('<CardAdd>', () => {
  let store;
  let props;
  let state;
  let component;
  let mockAddCard = jest.fn();

  beforeEach(() => {
    store = mockStore({});
    store.dispatch = mockAddCard;
    props = {
      listId: 'ListID',
    };
    state = {
      isOpened: false,
      title: '',
    };

    const wrapper = shallow(<CardAdd store={store} {...props} />);
    component = wrapper.find('CardAdd').dive();
    component.setState({ ...state })
  });

  it('should render AddButton if isOpened is false', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render form if isOpened is true', () => {
    component.instance().toggleOpened();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('checks if toggleOpened sets isOpened correctly', () => {
    component.instance().toggleOpened();
    expect(component.instance().state.isOpened).toEqual(true);

    component.instance().toggleOpened();
    expect(component.instance().state.isOpened).toEqual(false);
  });

  it("should call toggleOpened after click on AddButton", () => {
    const spy = jest.spyOn(component.instance(), "toggleOpened");

    component.find('AddButton').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should call toggleOpened after click on CancelButton", () => {
    const spy = jest.spyOn(component.instance(), "toggleOpened");

    component.setState({ isOpened: true });
    component.find('CancelButton').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });

  describe('add card form', () => {
    beforeEach(() => {
      component.setState({ isOpened: true });
    });

    it("should call toggleOpened if key code equal 27", () => {
      const spy = jest.spyOn(component.instance(), "toggleOpened");

      component.instance().handleKeyDown({ keyCode: 27 });

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call handleChange on title change with the correct params', () => {
      const spy = jest.spyOn(component.instance(), "handleChange");
      const mockEvent = {
        target: {
          name: "title",
          value: "Test title"
        }
      };

      component.find("TextArea").simulate("change", mockEvent);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(mockEvent);
    });

    it('should call setState on title', () => {
      const mockEvent = {
        target: {
          name: "title",
          value: "Test Title"
        }
      };
      const expectedState = {
        title: "Test Title",
        isOpened: true
      };

      component.instance().handleChange(mockEvent);

      expect(component.state()).toEqual(expectedState)
    });

    it("should call handleSubmit with the correct params", () => {
      const spy = jest.spyOn(component.instance(), "handleSubmit");
      const mockEvent = {
        preventDefault: jest.fn()
      };

      component.setState({ title: "Test Title" });
      component.instance().handleSubmit(mockEvent);

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it("should return handleSubmit if title is empty", () => {
      const spy = jest.spyOn(component.instance(), "handleSubmit");
      const mockEvent = {
        preventDefault: jest.fn()
      };

      component.instance().handleSubmit(mockEvent);

      expect(spy).toReturn();
    });

    it("should dispatch an action after handleSubmit", () => {
      const expectedAction = {
        type: 'ADD_CARD',
        payload: {
          listId: 'ListID',
          cardTitle: 'Test Title',
          newCardId: expect.any(String), //Date.now() id
        }
      };

      component.setState({ title: "Test Title" });
      component.find('SubmitButton').simulate('click');

      expect(mockAddCard).toHaveBeenCalledTimes(1);
      expect(mockAddCard).toHaveBeenCalledWith(expectedAction);
    });
  })
});

