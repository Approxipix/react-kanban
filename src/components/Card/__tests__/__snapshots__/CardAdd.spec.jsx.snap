// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardAdd> should render AddButton if isOpened is false 1`] = `
<AddButton
  onClick={[Function]}
>
  + Add another card
</AddButton>
`;

exports[`<CardAdd> should render form if isOpened is true 1`] = `
<ClickOutside
  action="mousedown"
  toggleOpened={[Function]}
>
  <styled.div>
    <form
      onSubmit={[Function]}
    >
      <TextArea
        autoFocus={true}
        name="title"
        onChange={[Function]}
        onKeyDown={[Function]}
        padding=".5rem"
        placeholder="Enter a card title..."
        rows={3}
        size=".875rem;"
        spellCheck={false}
        type="text"
        value=""
      />
      <styled.div>
        <SubmitButton
          disabled={true}
          type="submit"
        >
          Create card
        </SubmitButton>
        <CancelButton
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="times"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </CancelButton>
      </styled.div>
    </form>
  </styled.div>
</ClickOutside>
`;
